<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter. Выполнил - Артиков Артур
   

In [3]:
import pandas as pd
import psycopg2

In [88]:
DBNAME = 'confidential_dbname' # данные являются конфиденциальными и не могут быть вылажены в открытый доступ
USER = 'confidential_user'
PASSWORD = 'confidential_password'
HOST = 'confidential_host'
PORT = 'confidential_port'

In [5]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [89]:
!pip freeze > requirements.txt

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies).

In [6]:
# текст запроса
query_3_1 = f'''select count(*) /*используем count(*) чтобы подсчитать общее число вакансий*/
            from public.vacancies
            '''

In [7]:
count_of_vacancies = pd.read_sql_query(query_3_1, connection) # Получаем Датафрейм, состоящий из 1 элемента. Для вывода числа воспользуемся DF.iloc
print('Общее число вакансий -', count_of_vacancies.iloc[0]['count'])

<ipython-input-7-e824c5e14b8c>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_of_vacancies = pd.read_sql_query(query_3_1, connection) # Получаем Датафрейм, состоящий из 1 элемента. Для вывода числа воспользуемся DF.iloc


Общее число вакансий - 49197


2. Напишите запрос, который посчитает количество работодателей (таблица employers).

In [8]:
query_3_2 = f'''select count(*) /*используем count(*) чтобы подсчитать общее число работодателей*/
            from public.employers
            '''

In [9]:
count_of_employers = pd.read_sql_query(query_3_2, connection) # Получаем Датафрейм, состоящий из 1 элемента. Для вывода числа воспользуемся DF.iloc
print('Общее число работодателей -', count_of_employers.iloc[0]['count'])


Общее число работодателей - 23501


<ipython-input-9-e2f3aa17408b>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_of_employers = pd.read_sql_query(query_3_2, connection) # Получаем Датафрейм, состоящий из 1 элемента. Для вывода числа воспользуемся DF.iloc


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [10]:
query_3_3 = f'''
            select count(*) /*используем count(*) чтобы подсчитать общее число регионов*/
            from public.areas
            '''

In [11]:
count_of_areas = pd.read_sql_query(query_3_3, connection) # Получаем Датафрейм, состоящий из 1 элемента. Для вывода числа воспользуемся DF.iloc
print('Общее число регионов -', count_of_areas.iloc[0]['count'])

Общее число регионов - 1362


<ipython-input-11-7662abc8636c>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_of_areas = pd.read_sql_query(query_3_3, connection) # Получаем Датафрейм, состоящий из 1 элемента. Для вывода числа воспользуемся DF.iloc


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [12]:
query_3_4 = f'''
            select count(*) /*используем count(*) чтобы подсчитать общее число сфер деятельности*/
            from public.industries
            '''

In [13]:
count_of_industries = pd.read_sql_query(query_3_4, connection)
print('Общее число сфер деятельности -', count_of_industries.iloc[0]['count'])

Общее число сфер деятельности - 294


<ipython-input-13-19e12afe974d>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_of_industries = pd.read_sql_query(query_3_4, connection)


***

In [14]:
print('Общее число вакансий -', count_of_vacancies.iloc[0]['count'])
print('Общее число работодателей -', count_of_employers.iloc[0]['count'])
print('Общее число регионов -', count_of_areas.iloc[0]['count'])
print('Общее число сфер деятельности -', count_of_industries.iloc[0]['count'])

Общее число вакансий - 49197
Общее число работодателей - 23501
Общее число регионов - 1362
Общее число сфер деятельности - 294


### Выводы по предварительному анализу данных

На основании полученной информации, можно сделать несколько базовых выводов о составе таблиц. Во-первых, наибольшее число данных представленно в таблице с вакансиями, что доказывает верность того, что именно эта таблица взята в анализе как основная.

Число работодателей примерно в 2 раза меньше, чем число вакансий. Можно сделать вывод, что каждый из работодателей разместил в среднем 2-3 вакансии (конечно реальные данные могут отличаться, т.к. точное количество не было высчитано).

Число регионов, равное 1362 свидетельствует о большом географическом разбросе вакансий. Возможно также, что в данных есть информация о некоторых вакансиях не на территории России, однако для доказательства этого предположения нужно более детально изучить таблицу areas.

Наконец, наименьшим по числу является количество представленных сфер деятельности - всего 294. Можно предположить, что из-за отличия в написании сфер деятельности, при помощи разных регистров, или разных языков (английский, русский) число сфер деятельности получилось большим, нежели уникальное число таковых на самом деле. В ходе анализа следует дополнительно учесть данное обстоятельство, поскольку различные способы написания сфер деятельности стоит учитывать при поиске интересующего варианта для соискателя.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [15]:
query_4_1 = f'''
            select count(*) as "cnt", a.name as "areas"
            from public.vacancies as v
              join public.areas as a on v.area_id = a.id
            group by a.name
            order by cnt desc
            '''
# Объединяем наши таблицы по столбцам с идентичной информацией.

In [16]:
vacancies_areas = pd.read_sql_query(query_4_1, connection)
display(vacancies_areas.head(5))  # Выводим лидеров по числу вакансий

<ipython-input-16-b33be77bde23>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vacancies_areas = pd.read_sql_query(query_4_1, connection)


cnt            areas
0  5333           Москва
1  2851  Санкт-Петербург
2  2112            Минск
3  2006      Новосибирск
4  1892           Алматы

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [17]:
query_4_2 = f'''
            select count(*)
            from public.vacancies as v
            where v.salary_from is not null or v.salary_to is not null
            '''

In [18]:
completed_salary = pd.read_sql_query(query_4_2, connection)
print('Число вакансий, в которых заполнено хотя бы одно поле зарплаты -', completed_salary.iloc[0]['count'])

Число вакансий, в которых заполнено хотя бы одно поле зарплаты - 24073


<ipython-input-18-8f48c862835d>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  completed_salary = pd.read_sql_query(query_4_2, connection)


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [19]:
query_4_3 = f'''
            select round(avg(v.salary_from)) as "Нижняя граница зарплатной вилки", round(avg(v.salary_to)) as "Верхняя граница зарплатной вилки"
            from public.vacancies as v
            '''

In [20]:
salary_mean = pd.read_sql_query(query_4_3, connection)
print('Средняя нижняя граница зарплатной вилки, округленная до целого, равна -', salary_mean.iloc[0]['Нижняя граница зарплатной вилки'] , 'рублей')
print('Средняя верхняя граница зарплатной вилки, округленная до целого, равна -', salary_mean.iloc[0]['Верхняя граница зарплатной вилки'] , 'рублей')

Средняя нижняя граница зарплатной вилки, округленная до целого, равна - 71065.0 рублей
Средняя верхняя граница зарплатной вилки, округленная до целого, равна - 110537.0 рублей


<ipython-input-20-615c01aa95b5>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  salary_mean = pd.read_sql_query(query_4_3, connection)


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [21]:
query_4_4 = f'''
            select v.schedule as "Тип рабочего графика", v.employment as "Тип трудоустройства", count(*) as "Количество вакансий"
            from public.vacancies as v
            group by v.schedule, v.employment
            order by "Количество вакансий" desc
            '''

In [22]:
schedule_employment = pd.read_sql_query(query_4_4, connection)
display(schedule_employment.head()) # Верхняя часть созданного Датафрейма
print('Второе по популярности сочетание рабочего графика и типа трудоустроства:', schedule_employment.iloc[1], sep='\n')

<ipython-input-22-d8e85c57843b>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  schedule_employment = pd.read_sql_query(query_4_4, connection)


Тип рабочего графика  Тип трудоустройства  Количество вакансий
0          Полный день     Полная занятость                35367
1     Удаленная работа     Полная занятость                 7802
2        Гибкий график     Полная занятость                 1593
3     Удаленная работа  Частичная занятость                 1312
4       Сменный график     Полная занятость                  940

Второе по популярности сочетание рабочего графика и типа трудоустроства:
Тип рабочего графика    Удаленная работа
Тип трудоустройства     Полная занятость
Количество вакансий                 7802
Name: 1, dtype: object


5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта.

In [23]:
query_4_5 = f'''
            select v.experience as "Требуемый опыт работы", count(*) as "Число вакансий"
            from public.vacancies as v
            group by v.experience
            order by  "Число вакансий" asc

            '''

In [24]:
experience_count = pd.read_sql_query(query_4_5, connection)
display(experience_count)

<ipython-input-24-763d9372d4da>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  experience_count = pd.read_sql_query(query_4_5, connection)


Требуемый опыт работы  Число вакансий
0           Более 6 лет            1337
1             Нет опыта            7197
2         От 3 до 6 лет           14511
3    От 1 года до 3 лет           26152

***

### Выводы по детальному анализу вакансий

Во время решения задания 4.1 были получены полезные сведения о городах, с наибольшим числом размещенных вакансий. В топ-5 городов по числу вакансий вошли топ-3 города России по населению (Москва, Санкт-Петербург и Новосибирск), а также два города с наибольшим число жителей Казахстана (Алматы) и Республики Беларусь (Минск). Также получило подтверждение теория, выдвинутая по выводам предварительного анализа данных, а именно - наличие в базе вакансий не только России, но и других стран, таких как Казахстан и Беларусь.

При решении задания 4.2 было найдено число вакансий, в которых заполнено хотя бы одно поле с заработной платой. Число таковых вакансий - 24073, то есть менее 50% от числа всех вакансий. Графа с заработной платой очень сильно влияет на желание соискателя пойти на собеседование на ту, или иную должность, поскольку отсутствие четких рамок, или хотя бы четкой нижней (верхней) границы в зарплатах может служить триггером для отказа от собеседования, в пользу той компании, где данная граница записана, ввиду страха быть обманутым, или нежелания соискателя идти туда, где платят неизвестное количество денег.

Задание 4.3 помогло понять, как в среднем выглядит зарплатная вилка для вакансий. В среднем соискатель может рассчитывать на заработную плату в размере от 71065 рублей, до 110537 рублей. Эти данные могут помочь в будущем при отбрасывании значений, которые являются экстремально большими, или мелкими по сравнению с нормальной зарплатной вилкой.

Предпоследнее задание дало нам возможность понять, какого соотношение типа рабочего графика и типа трудоустройства и количество вакансий по данным соотношениям. Наиболее популярным является сочетание "полный день - полная занятость", на втором месте "удаленная работа - полная занятость" и замыкает тройку лидеров "гибкий график - полная занятость". Как видно, во всех вариантах присутствует полная занятость, следовательно у соискателя больше шансов найти работу, если он готов к работе в стандартном режиме 5/2 по 8 часов в день, чем при других вариантах времени работы.

Наконец, решение задания 4.5 дало нам возможность понять, какого число вакансий по требуемому опыту работы. Больше всего вакансий, подразумевают наличие опыта "от 1 года до 3 лет", в 2 раза меньше вакансий требуют опыт "от 3 до 6 лет", что соответствует уровню middle-senior, на предпоследнем месте вакансии, не требующие опыта работы. Наименьшее число представленных вакансий предполагает наличие опыта от 6 лет и более, что соответствует уровню senior-lead. Данное распределение достаточно легко понять. Более всего интересны соискатели с уровнями junior+ - middle+, вакансии, требующие большего опыта достаточно редки, поскольку наличие рабочих мест (позиций в компании), требующих такого большого опыта - крайне невелико.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [25]:
query_5_1 = f'''
            select distinct e.name as "Название компании", count(*) as "Число вакансий"
            from public.vacancies as v
                 join public.employers as e on v.employer_id = e.id
            group by e.name
            order by "Число вакансий" desc
            '''

In [26]:
employer_count = pd.read_sql_query(query_5_1, connection)
print('Работадатель, на первом месте по количеству вакансий -', employer_count.iloc[0]['Название компании'])
print('Работодатель, на пятом месте по количеству вакансий -', employer_count.iloc[4]['Название компании'])

<ipython-input-26-30fe35558223>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employer_count = pd.read_sql_query(query_5_1, connection)


Работадатель, на первом месте по количеству вакансий - Яндекс
Работодатель, на пятом месте по количеству вакансий - Газпром нефть


2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [28]:
query_5_2 = f'''
            select a.name as "Название региона", count(v.id) as "Количество вакансий", count(e.id) as "Количество работодателей"
            from public.areas as a
                 left join public.vacancies as v on a.id = v.area_id
                 left join public.employers as e on a.id = e.area
            where v.id is null
            group by a.name
            order by  "Количество работодателей" desc
            '''

In [29]:
employers_nullvacancies = pd.read_sql_query(query_5_2, connection)
display(employers_nullvacancies.head())
print('\nРегион с наибольшим числом работодателей, не имеющий вакансий -', employers_nullvacancies.iloc[0]['Название региона'])

<ipython-input-29-af56aa34da58>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_nullvacancies = pd.read_sql_query(query_5_2, connection)


Название региона  Количество вакансий  Количество работодателей
0              Россия                    0                       410
1           Казахстан                    0                       207
2  Московская область                    0                        75
3  Краснодарский край                    0                        19
4            Беларусь                    0                        18


Регион с наибольшим числом работодателей, не имеющий вакансий - Россия


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [30]:
query_5_3 = f'''
select distinct e.name as "Название компании", count(distinct v.area_id) as "Количество регионов"
from public.vacancies as v
     left join public.employers as e on v.employer_id = e.id
group by e.name
order by "Количество регионов" desc
  '''

In [31]:
employer_countareas = pd.read_sql_query(query_5_3, connection)
display(employer_countareas.head())
print('Компания', employer_countareas.iloc[0]["Название компании"], 'имеет вакансии в', employer_countareas.iloc[0]["Количество регионов"], 'различном регионе')

<ipython-input-31-43bf9bd140dd>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employer_countareas = pd.read_sql_query(query_5_3, connection)


Название компании  Количество регионов
0                  Яндекс                  181
1              Ростелеком                  152
2              Спецремонт                  116
3  Поляков Денис Иванович                   88
4               ООО ЕФИН                    71

Компания Яндекс имеет вакансии в 181 различном регионе


4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности.

In [32]:
query_5_4 = f'''
select count(*) as "Число работодателей" /*Число работодателей с неуказанными сферами деятельности*/
from public.employers as e
     left join public.employers_industries as i on e.id = i.employer_id
where i.industry_id is null
'''

In [33]:
employers_nullindustries = pd.read_sql_query(query_5_4, connection)
print('Число работодателей с неуказанными сферами деятельности', employers_nullindustries.iloc[0]['Число работодателей'])

Число работодателей с неуказанными сферами деятельности 8419


<ipython-input-33-e7bb1db44e72>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_nullindustries = pd.read_sql_query(query_5_4, connection)


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности.

In [34]:
query_5_5 = f'''
select e.name "Название компании"
from public.employers as e
     left join public.employers_industries as i on e.id = i.employer_id
group by e.name
having count(*) = 4
order by e.name asc
'''

In [35]:
company_4industries = pd.read_sql_query(query_5_5, connection)
display(company_4industries.head())
print('Третья в алфавитном порядке компания, имеющая 4 сферы деятельности -', company_4industries.iloc[2]['Название компании'])

<ipython-input-35-24f76cfca1ef>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  company_4industries = pd.read_sql_query(query_5_5, connection)


Название компании
0      101 Интернет
1          21vek.by
2              2ГИС
3                2К
4       4 пикселя +

Третья в алфавитном порядке компания, имеющая 4 сферы деятельности - 2ГИС


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [36]:
query_5_6 = f'''
select count(*)
from public.employers_industries as ei
     join public.industries as i on ei.industry_id = i.id
where name = 'Разработка программного обеспечения'
'''

In [37]:
count_employers_software = pd.read_sql_query(query_5_6, connection)
print('Число компаний, занимающихся разработкой программного обеспечения -', count_employers_software.iloc[0]['count'])

Число компаний, занимающихся разработкой программного обеспечения - 3553


<ipython-input-37-e1e617cdc432>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_employers_software = pd.read_sql_query(query_5_6, connection)


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8).

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API.

In [38]:
import requests
from bs4 import BeautifulSoup
url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России'
response = requests.get(url)
page = BeautifulSoup(response.text, 'lxml')

cities_links = list()
for i in range(2, 18):
  link = f'#mw-content-text > div > table.standard.sortable > tbody > tr:nth-child({i}) > td:nth-child(2) > a' # создаем прообраз для селектора на ссылки на названия городов - миллионников
  cities_links.append(page.select(link))    # получаем текст ссылки, а также название города, внутри тега <a>

for i in range(len(cities_links)):
  cities_links[i] = str(*cities_links[i])   # переводим тип полученной ссылки с текстом в строковый вид

cities = list()
for city_link in cities_links:
  city = city_link.split('"')[-2]           # разделям строку по символам двоной кавычки, после чего получаем минус второй элемент - название города
  cities.append(city)

cities = tuple(cities)                      # переводим список с городами в кортеж


In [39]:
query_5_7 = f'''
select a.name as "Название города", count(*) as "Число вакансий"
from public.vacancies as v
     left join public.areas as a on v.area_id = a.id
     left join public.employers as e on v.employer_id = e.id
where a.name in {cities} and e.name = 'Яндекс'
group by a.name

union all

(select 'Total', count(*)
from public.vacancies as v
     left join public.areas as a on v.area_id = a.id
     left join public.employers as e on v.employer_id = e.id
where a.name in {cities} and e.name = 'Яндекс'
order by 2 desc
limit 1)

'''

In [40]:
yandex_vacancies = pd.read_sql_query(query_5_7, connection)
display(yandex_vacancies)

<ipython-input-40-d015f4803ecf>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  yandex_vacancies = pd.read_sql_query(query_5_7, connection)


Название города  Число вакансий
0         Волгоград              24
1           Воронеж              32
2      Екатеринбург              39
3            Казань              25
4         Краснодар              30
5        Красноярск              23
6            Москва              54
7   Нижний Новгород              36
8       Новосибирск              35
9              Омск              21
10            Пермь              25
11   Ростов-на-Дону              25
12           Самара              26
13  Санкт-Петербург              42
14              Уфа              26
15        Челябинск              22
16            Total             485

***

### Выводы по анализу работодателей

В ходе анализа работодателей удалось дополнительно прояснить, какие компании лучше всего выбирать для поиска в них работы, а также какие из работодателей наиболее заинтересованы в привлечении новых сотрудников.

Первое задание позволило нам узнать соотношение работодателей и числа предлагаемых ими вакансий. В топ-3 данного списка вошли Яндекс, Ростелеком и Тинькофф. Следовательно, при поиске работы следует обратить внимание на вакансии, предлагаемые данными организациями, поскольку наличие большого числа открытых вакансий свидетельствует о том, что компания развивает новые направления и продукты, а следовательно ей требуются новые специалисты.

Следующее задание позволило нам узнать те регионы, в которых нет вакансий, однако есть работодатели. Рекордсменом в данной категории стала "Россия", как отдельный регион. Это связано в первую очередь с тем, что некоторым работодателям не важно нахождение человека в конкретном городе, следовательно в вакансии указывается местоположение "Россия", как приемлемый регион (возможно, в данных компаниях делается акцент на удаленную работу).

Задание 5.3 позволило нам узнать о количестве регионов, в которых работодатель публикует свои вакансии. Чем больше компания - тем в большем числе регионов она будет публиковать свои предложения о работе, чтобы находить различных соискателей, готовых работать по разным направлениям.

При решении задания 5.4 выяснилось, что из представленных работодателей - 8419 работодателей не указывают сферы деятельности. Данная ситуация может происходить по одной из двух причин:
1. Род деятельности компании понятен из ее названия, например "МВД России по Республике Башкортостан"
2. Компания является частью более большого конгломерата, а вакансия публикуется в рамкой отдельной части данного конгломерата, например "2ГИС-Хабаровск"
Тем не менее, также есть компании, сфера деятельности которых непонятна, а название не дает каких-либо дополнительных представлений о области, в которой работает компания. Для соискателя отсутствие описанной сферы деятельности у компании может стать причиной для отказа от отклика на вакансию, поскольку навыки, необходимые для работы в компании часто зависят от сферы ее деятельности

Задание 5.5 и 5.6 расширили наше познание в количествах сфер деятельности компаний, а также в видах данных сфер деятельности. Существует огромное число компаний, работающий более чем в 1 сфере. Это опять-таки свидетельствует о том, что компания развивает разносторонние продукты и является активным участником рынка.

Финальное задание позволило лучше ознакомиться с одним из лидеров рынка - компанией Яндекс, а также с тем, какого число вакансий у данной компании, в городах миллионниках.



# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.*


In [41]:
query_6_1 = f'''
select count(*)
from public.vacancies as v
where lower(v.name) like '%data%' or lower(v.name) like '%данн%'
'''

In [42]:
data_vacancies = pd.read_sql_query(query_6_1, connection)
print('Число вакансий имеющих отношение к данным -', data_vacancies.iloc[0]['count'])

<ipython-input-42-80da9b0526b8>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data_vacancies = pd.read_sql_query(query_6_1, connection)


Число вакансий имеющих отношение к данным - 1771


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста?
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.


In [43]:
query_6_2 = f'''
select count(*)
from public.vacancies as v
where (lower(v.name) like '%data scientist%' or
       lower(v.name) like '%data science%' or
       lower(v.name) like '%исследователь данных%' or
       (v.name like '%ML%' and v.name not like '%HTML%') or
       lower(v.name) like '%machine learning%' or
       lower(v.name) like '%машинн%обучен%')
       and
       (lower(v.name) like '%junior%' or
       v.experience like 'Нет опыта' or
       v.employment like 'Стажировка')
'''

In [44]:
junior_ds_vacancies = pd.read_sql_query(query_6_2, connection)
print('Число вакансий для начинающих Data Scientist - специалистов:', junior_ds_vacancies.iloc[0]['count'])

<ipython-input-44-c8ae5a27c97b>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  junior_ds_vacancies = pd.read_sql_query(query_6_2, connection)


Число вакансий для начинающих Data Scientist - специалистов: 51


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [45]:
query_6_3 = f'''
select count(*)
from public.vacancies as v
where (lower(v.name) like '%data scientist%' or
       lower(v.name) like '%data science%' or
       lower(v.name) like '%исследователь данных%' or
       (v.name like '%ML%' and v.name not like '%HTML%') or
       lower(v.name) like '%machine learning%' or
       lower(v.name) like '%машинн%обучен%')
       and
       (lower(v.key_skills) like '%sql%' or
       lower(v.key_skills) like '%postgres%')

except

select count(*)
from public.vacancies as v
where (lower(v.name) like '%data scientist%' or
       lower(v.name) like '%data science%' or
       lower(v.name) like '%исследователь данных%' or
       (v.name like '%ML%' and v.name not like '%HTML%') or
       lower(v.name) like '%machine learning%' or
       lower(v.name) like '%машинн%обучен%')
       and
       (lower(v.name) like '%junior%' or
       v.experience like 'Нет опыта' or
       v.employment like 'Стажировка')
'''

In [46]:
count_notjunior_ds = pd.read_sql_query(query_6_3, connection)
print('Число вакансий, для уровня выше Junior Data Scientis, с ключевыми навыками "SQL"/"Postgres" -', count_notjunior_ds.iloc[0]['count'])

<ipython-input-46-bc194b470660>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_notjunior_ds = pd.read_sql_query(query_6_3, connection)


Число вакансий, для уровня выше Junior Data Scientis, с ключевыми навыками "SQL"/"Postgres" - 201


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [47]:
query_6_4 = f'''
select count(*)
from public.vacancies as v
where (lower(v.name) like '%data scientist%' or
       lower(v.name) like '%data science%' or
       lower(v.name) like '%исследователь данных%' or
       (v.name like '%ML%' and v.name not like '%HTML%') or
       lower(v.name) like '%machine learning%' or
       lower(v.name) like '%машинн%обучен%')
       and
       (lower(v.key_skills) like '%python%')

except

select count(*)
from public.vacancies as v
where (lower(v.name) like '%data scientist%' or
       lower(v.name) like '%data science%' or
       lower(v.name) like '%исследователь данных%' or
       (v.name like '%ML%' and v.name not like '%HTML%') or
       lower(v.name) like '%machine learning%' or
       lower(v.name) like '%машинн%обучен%')
       and
       (lower(v.name) like '%junior%' or
       v.experience like 'Нет опыта' or
       v.employment like 'Стажировка')
'''

In [48]:
ds_python_keyskill = pd.read_sql_query(query_6_4, connection)
print('Число DS - вакансий, для уровня выше Junior, с ключевым навыком Python -', ds_python_keyskill.iloc[0]['count'])

<ipython-input-48-ebb4d13c118a>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  ds_python_keyskill = pd.read_sql_query(query_6_4, connection)


Число DS - вакансий, для уровня выше Junior, с ключевым навыком Python - 351


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [65]:
query_6_5 = f'''
select 1+length(v.key_skills) - length(replace(v.key_skills, chr(9), '')) as "count_of_keyskills"
from public.vacancies as v
where (lower(v.name) like '%data scientist%' or
       lower(v.name) like '%data science%' or
       lower(v.name) like '%исследователь данных%' or
       (v.name like '%ML%' and v.name not like '%HTML%') or
       lower(v.name) like '%machine learning%' or
       lower(v.name) like '%машинн%обучен%')
       and
       v.key_skills is not null
'''

In [66]:
key_skills_ds = pd.read_sql_query(query_6_5, connection)
print('В вакансиях для DS чаще всего указывают', round(key_skills_ds['count_of_keyskills'].mean(), 2), 'навыка')

<ipython-input-66-9200dd17debe>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  key_skills_ds = pd.read_sql_query(query_6_5, connection)


В вакансиях для DS чаще всего указывают 6.41 навыка


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*).

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа.

In [75]:
query_6_6 = f'''
select distinct v.experience, round(avg(coalesce((v.salary_from+v.salary_to)/2, salary_from, salary_to))) as salary
from public.vacancies as v
where (lower(v.name) like '%data scientist%' or
       lower(v.name) like '%data science%' or
       lower(v.name) like '%исследователь данных%' or
       (v.name like '%ML%' and v.name not like '%HTML%') or
       lower(v.name) like '%machine learning%' or
       lower(v.name) like '%машинн%обучен%')
       and
       (v.salary_from is not null or v.salary_to is not null)
group by v.experience
'''

In [76]:
ds_expierence_salary = pd.read_sql_query(query_6_6, connection)
print('Средняя заработная плата, на которую может рассчитывать Data Scientist с опытом работы от 3 до 6 лет -', ds_expierence_salary.iloc[2]['salary'])

<ipython-input-76-e1502e266b3f>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  ds_expierence_salary = pd.read_sql_query(query_6_6, connection)


Средняя заработная плата, на которую может рассчитывать Data Scientist с опытом работы от 3 до 6 лет - 243115.0


***

### Выводы по предметному анализу

После предметного анализа данных, были полученны новые данные, относительно вакансий, так или иначе связанных с Data Science.
* Во-первых, было получено число вакансий, так или иначе связанных с данными, их оказалось 1771.
* Во-вторых, среди всех вакансий были найдены только те, что подходят для начинающего Data Science - специалиста. За ключевые признаки были взяты слова "junior" в названии вакансии, "Нет опыта" - в графе требуемого опыта работы и "Стажировка", как тип трудоустройства.
* Следующим шагом бало нахождение вакансий, с указанными требованиями, такими как SQL и Postgres (таких вакансий представлено 201) и Python (таких вакансий представлено 351) для уровня выше Junior (задания 3-4).
* Пятым шагом был подсчет среднего числа ключевых навыков, для вакансий
* Последним шагом была подсчитана заработная плата, на которую может рассчитывать специалист с разным опытом. При этом отбрасывались вакансии, в которых не была указана ни нижняя, ни верхняя граница заработной платы.

# Дополнительные исследования

1. В ходе решения мы посчитали зависимость заработной платы от опыта, а также среднее значение для минимального и максимального значения заработной платы. Теперь посчитаем среднюю заработную плату по городам

In [77]:
extra_query_1 = f'''
select a.name as "Область", round(avg(coalesce((v.salary_from+v.salary_to)/2, salary_from, salary_to))) as "Средняя З/П по городу"
from public.vacancies as v
    left join public.areas as a on v.area_id = a.id
where (v.salary_from is not null or v.salary_to is not null)
group by a.name
order by "Средняя З/П по городу" desc
'''

In [79]:
salary_areas = pd.read_sql_query(extra_query_1, connection)
display(salary_areas.head())

<ipython-input-79-a02ff90ed792>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  salary_areas = pd.read_sql_query(extra_query_1, connection)


Область  Средняя З/П по городу
0    Германия               452751.0
1    Болгария               324348.0
2     Испания               300000.0
3  Черногория               288589.0
4       Литва               263019.0

### Вывод по дополнительному исследованию

Как видно из полученного Датафрейма, наибольшая средняя зарплата, округленная до целого числа, среди всех представленных областей возможно заработать в Германии. Следующие страны, находящиеся в топ-5, также относятся к Евросоюзу, это: Германия, Болгария, Испания, Черногория и Литва. На основании собранных данных можно сделать о том, что средяя зарплата в Европе в средне выше, чем в России и странах-членах СНГ.

2. Теперь получим топ компаний, по числу их сфер деятельности


In [83]:
extra_query_2 = f'''
select distinct e.id, e.name, count(*) as "Число сфер деятельности"
from public.employers as e
     left join public.employers_industries as ei on e.id = ei.employer_id
     join public.industries as i on ei.industry_id = i.id
group by e.id, e.name
order by "Число сфер деятельности" desc
'''

In [86]:
count_industries = pd.read_sql_query(extra_query_2, connection)
display(count_industries.head(10))

<ipython-input-86-b77dc2cae596>:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_industries = pd.read_sql_query(extra_query_2, connection)


id                                 name  Число сфер деятельности
0   15                          Наука-Связь                        5
1  205                                 Марс                        5
2  228        GE (General Electric Company)                        5
3  272                          DHL Express                        5
4  330                               Nestle                        5
5  355                   Philips HealthTech                        5
6  389  БИОКАД, биотехнологическая компания                        5
7  540                                ZEBRA                        5
8  616                        Lanck Telecom                        5
9  673                   METRO Cash & Carry                        5

### Вывод по дополнительным исследованиям

Как видно из полученных данных, почти все компании, имеющие 5 сфер деятельности являются транснациональными корпорациями. В топ-10 компаний вошли General Electric, DHL Express, Nestle, Philips, METRO, Zebra и Lanck Telekom. Также в числе лидеров такие компании как Биокад - крупнейшая Российская биотехнологическая компания полного цикла, Наука-Связь - Российская компания, предоставляющая телекоммуникационные услуги (телефония, интернет и прочее), а также Марс - один из мировых лидеров по производству пищевых продуктов длительного хранения (производит шоколадные батончики, корма для животных и прочее).

В целом, можно сделать вывод о том, что от размера компании зависит то, какого число ее сфер деятельности. Чем больше компания, тем больше у нее сфер деятельности, следовательно тем больше возможности у соискателя попасть в эту компанию на работу.

In [87]:
connection.close()

# Общий вывод по проекту

В ходе работы над проектом были получены выводы, которые могут помочь соискателю в поиске работы, а также помочь работодателю грамотнее искать исполнителей на свои вакансии.

*Что будет полезно для соискателя:*
1. Наибольшее число вакансий возможно найти в больших городах (в полученном исследовании это были Москва, Санкт-Петербург, Новосибирск, Минск и Алматы). Следовательно, при поиске работы лучше находиться в крупном городе, или быть готовым к переезду в один из вышеперечисленных крупных городов
2. Средняя зарплатная вилка среди всех вакансий имеет вид: 71-110 тыс. рублей. При поиске работы нужно обращать внимание на данную вилку, поскольку для начинающих специалистов ориентировка на нее способна помочь мыслить критически и понимать, какие зарплаты стоит или не стоит ожидать.
3. Почти все работодатели заинтересованы в работе сотрудника в формате полной занятости. График может варьироваться от гибридного до удаленного, однако нужно быть готовым работать стандартные 8 часов 4-5 дней в неделю, если соискатель хочет расти в карьере и получать достойную зарплату.
4. Опыт работы влияет на количество вакансий. Наибольшее число вакансий включает в себя требование к опыту: от 1 года до 3 лет, следовательно наличие опыта в той сфере, в которой работает компания станет плюсом как при поиске вакансии, так и при обсуждении размера заработной платы.
5. Поиск работы в крупных компаниях, таких как "Яндекс", "Тинькофф" и "Ростелеком" с б__о__льшей вероятностью приведут к успеху, поскольку наличие большого числа вакансий дает соискателю шанс выбрать похожее направление работы, при неудаче в одном из таковых.
6. Следует обращать внимание на сферу деятельности работодателя, чтобы лучше понимать свои силы. Познания в области программирования, направленные на медицинску сферу навряд ли будут также эффективны при работе в банке. Именно соответствие сферы работы компании умениям соискателя приведут в конечном итоге к выгоде для обеих сторон.
7. Для соискателей уровня выше Middle и выше, имеется возможность найти работу за границей. Тем не менее стоит также учитывать то, что для работы за границей потребуется дополнительное знание как минимум двух языков: местного языка (для Германии - немецкий, для Испании - испанский, и.т.д), а также английского, чтобы иметь возможность работать с документацией и прояснять рабочие моменты.
8. Для специалистов в сфере Data Science на рынке присутствует огромное число вакансий всех возможных вариаций. Начинающие специалисты в данной сфере также имеют возможность попасть на стажировку/работу, чтобы, в случае ее удачного прохождения, получить оффер.
9. Следует обращать внимание на ключевые навыки, указанные компанией. Соответствие ключевым навыкам работадателя гарантированно поможет в получении работы, а также поможет понять, когда необходимо получить новые знания.

*Что будет полезно для работодателей*
1. Географический разброс вакансий. Если компания обладает разными вакансиями, в разных частях страны, или даже мира, то это с высокой вероятностью привет к тому, что компания легко сможет найти нового работника, по совершенно разным проектам.
2. Наличие удаленного формата работы. Компании, дающие возможность своим работникам работать как удаленно, так и из офиса будут иметь возможность нанять почти любого сотрудника. Также наличие удаленного формата работы, помимо работы в течение полного дня, создаст у работника ощущение доверия и свободы, поскольку незапланированная поездка, или форс-мажорная ситуация никак негативно не скажется ни на работе сотрудника, ни на решении проекта компании.
3. Необходимо указывать сферу деятельности компании. Те фирмы, которые не указывают свои сферы деятельности могут испытывать трудности с поиском персонала. Очень часто соискатель обращает внимание на сферу деятельности компании, в котороую он будет подавать резюме. Отсутствие данной информации может привести к тому, что соискатель просто выберет другую компанию для работы.
4. Необходимо указывать зарплатную вилку (желательно обе границы). В огромном количестве случаев соискатель ищет работу, основываясь на желании получать определенную заработную плату. Если у вакансии отсуствуют четкие границы для ЗП, то соискатель с меньшей вероятностью откликнется на нее, следовательно компании придется дольше искать работника на определенную должность.
5. Ключевые навыки - главный фактор для соискателя. Наличие описанных ключевых навыков дают возможность будущему работнику развивать те навыки, которых у него пока что нет, а также понимать, соответсвтует ли он необходимым условиям для получения работы в данной компании.
